src/types.rs

```diff
@@ -1086,6 +1087,129 @@ impl SimpleAsn1Writable for GeneralizedTime {
     }
 }
 
+/// Used for parsing and writing ASN.1 `GENERALIZED TIME` values accepting
+/// fractional seconds value.
+/// See https://github.com/alex/rust-asn1/issues/491 for discussion.
+#[derive(Debug, Clone, PartialEq, PartialOrd, Hash, Eq)]
+pub struct GeneralizedTimeFractional {
+    pub datetime: DateTime,
+    pub fraction: Option<u32>, // Up to 1 ns precision (10^9)
+}
+
+impl GeneralizedTimeFractional {
+    pub fn new(dt: DateTime, fraction: Option<u32>) -> ParseResult<GeneralizedTimeFractional> {
+        Ok(GeneralizedTimeFractional {
+            datetime: dt,
+            fraction,
+        })
+    }
+
+    pub fn as_datetime(&self) -> &DateTime {
+        &self.datetime
+    }
+
+    pub fn fraction(&self) -> Option<u32> {
+        // Fractional time: between 1 and 10^9 (nanoseconds)
+        self.fraction
+    }
+}
+
+fn read_byte_no_advance(data: &[u8]) -> ParseResult<u8> {
+    if data.is_empty() {
+        return Err(ParseError::new(ParseErrorKind::InvalidValue));
+    }
+
+    Ok(data[0])
+}
+
+fn read_fractional_time(data: &mut &[u8]) -> ParseResult<Option<u32>> {
+    // We cannot use read_byte here because it will advance the pointer
+    // However, we know that the is suffixed by 'Z' so reading into an empty
+    // data should lead to an error.
+    if read_byte_no_advance(data)? == b'.' {
```

Comment:
I think it'd be simpler to just do `if data.first() == Some(b'.')` and then allow the _next_ caller (looking for the mandatory `Z`) to error out.

In general I think that style of error handling is much simpler.

Reply:
I've rewrote it using `.first()` and removed the `read_byte_no_advance` helper.